spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -194,19 +213,32 @@ protected MessageHandler createProducerMessageHandler(final ProducerDestination
 		if (errorChannel != null) {
 			handler.setSendFailureChannel(errorChannel);
 		}
-		String[] headerPatterns = producerProperties.getExtension().getHeaderPatterns();
-		if (headerPatterns != null && headerPatterns.length > 0) {
-			List<String> patterns = new LinkedList<>(Arrays.asList(headerPatterns));
-			if (!patterns.contains("!" + MessageHeaders.TIMESTAMP)) {
-				patterns.add(0, "!" + MessageHeaders.TIMESTAMP);
+		KafkaHeaderMapper mapper;
+		if (this.configurationProperties.getHeaderMapperBeanName() != null) {
+			mapper = getApplicationContext().getBean(this.configurationProperties.getHeaderMapperBeanName(),
+					KafkaHeaderMapper.class);
+		}
+		if (producerProperties.getHeaderMode() != null
```

Comment:
`else if()` ?

Because it looks now like we always override it here or in the next `else`

Reply:
That was my intent - even if the user provides a bean, if the header mode is `raw` ,`none` or `embeddedHeaders` we need to signal to the channel adapter that it should not populate headers.

Reply:
But I see the else needs help to not overwrite.